File: WalletTracker/app/src/main/java/com/chtima/wallettracker/fragments/HomeFragment.java
package com.chtima.wallettracker.fragments;

import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageButton;

import com.chtima.wallettracker.MainActivity;
import com.chtima.wallettracker.R;
import com.chtima.wallettracker.adapters.TransactionAdapter;
import com.chtima.wallettracker.components.Swicher;
import com.chtima.wallettracker.models.CategoryWithTransactions;
import com.chtima.wallettracker.models.DialogObserver;
import com.chtima.wallettracker.models.Transaction;
import com.chtima.wallettracker.models.TransactionType;
import com.chtima.wallettracker.models.User;
import com.chtima.wallettracker.vm.CategoryViewModel;
import com.chtima.wallettracker.vm.TransactionViewModel;
import com.chtima.wallettracker.vm.UserViewModel;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.PieEntry;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collection;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

import autodispose2.AutoDispose;
import autodispose2.androidx.lifecycle.AndroidLifecycleScopeProvider;
import io.reactivex.rxjava3.disposables.CompositeDisposable;

public class HomeFragment extends Fragment {

    private User user = null;

    private TransactionType transactionType; // transactionType is used in "filterTransactionWithUpdateUI"

    private final List<CategoryWithTransactions> categoryWithTransactions = new ArrayList<>();


    //ViewModels
    private TransactionViewModel transactionVM;
    private CategoryViewModel categoryVM;
    private UserViewModel userVM;

    //ui
    private RecyclerView recyclerView;

    private Button btnBalance;

    private Swicher swicher;

    private SliderChartFragment sliderChartFragment;

    //Adapters
    private TransactionAdapter transactionAdapter;
    private AddTransactionDialogFragment addTransactionDialogFragment;

    private HomeFragment(){}

    public static HomeFragment newInstance() {
        return new HomeFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        categoryVM = new ViewModelProvider(this).get(CategoryViewModel.class);
        transactionVM = new ViewModelProvider(this).get(TransactionViewModel.class);
        userVM = new ViewModelProvider(this).get(UserViewModel.class);

        updateCategoriesWithTransactions(); //only once!
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_home, container, false);

        //ui
        ((ImageButton)view.findViewById(R.id.btn_add)).setOnClickListener(x -> {

            addTransactionDialogFragment = AddTransactionDialogFragment.newInstance();
            addTransactionDialogFragment.setSubscribe(new DialogObserver<Transaction>() {
                @Override
                public void onSuccess(Transaction obj) {
                   transactionVM.insert(obj)
                           .to(AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(HomeFragment.this)))
                           .subscribe(id -> {
                               obj.id = id;
                               if (TransactionType.EXPENSE == obj.type)
                                   user.deductFromBalance(obj.sum);
                               else
                                   user.addToBalance(obj.sum);

                               btnBalance.setText(String.valueOf(user.balance));

                               userVM.update(HomeFragment.this.user)
                                       .to(AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(HomeFragment.this)))
                                       .subscribe(
                                               () -> Log.i("WW-INFO-C", "OK"),
                                               e -> Log.e("WW-INFO-E", e.getMessage())
                                       );

                               //updateCategoriesWithTransactions();
                           }, ex -> Log.e("ERR", ex.getMessage()));
                }

                @Override
                public void onCancel() {}
            });
            addTransactionDialogFragment.show(getChildFragmentManager(), AddTransactionDialogFragment.class.getName());
        });

        //recyclerView
        transactionAdapter = new TransactionAdapter(getContext(), toTransactionList());
        recyclerView = view.findViewById(R.id.transaction_recycle);
        recyclerView.setAdapter(transactionAdapter);

        sliderChartFragment = SliderChartFragment.newInstance();
        getChildFragmentManager().beginTransaction()
                .replace(R.id.slide_chart_fragment, sliderChartFragment, SliderChartFragment.class.getName())
                .commit();

        //swicher_transaction_type
        swicher = view.findViewById(R.id.swicher_transaction_type);
        swicher.setOnChangedSelectionListener(transactionType -> {
            this.transactionType = transactionType;
            this.filterTransactionWithUpdateUI();
        });

        //btns
        btnBalance = view.findViewById(R.id.balance_btn);
        btnBalance.setText(user.balance + "$");//set user balance to "balance_btn"

        return view;
    }

    //Method to filter transactions based on the transaction type and update the UI
    private void filterTransactionWithUpdateUI(){
        if(categoryWithTransactions.isEmpty()) return;

        List<Transaction> transactionsFilterType = categoryWithTransactions.stream()
                .map(x -> x.transactions)
                .flatMap(Collection::stream)
                .filter(x -> x.type == HomeFragment.this.transactionType)
                .sorted(Comparator.comparing((Transaction s) -> s.dateTime).reversed())
                .collect(Collectors.toList());

        transactionAdapter.updateList(transactionsFilterType);

        //setPieChartToday
        List<PieEntry> pieChartToday = new ArrayList<>();
        categoryWithTransactions.stream().map(item -> {
            double sum = item.transactions.stream()
                    .filter(
                            x -> x.type == HomeFragment.this.transactionType &&
                                    x.dateTime.toInstant()
                                            .atZone(ZoneId.systemDefault())
                                            .toLocalDate().equals(
                                                    MainActivity.nowDate.toInstant().atZone(ZoneId.systemDefault()).toLocalDate()
                                            )
                    )
                    .mapToDouble(t -> t.sum)
                    .sum();
            return new AbstractMap.SimpleEntry<>(item, sum);
        }).filter(x -> x.getValue() > 0)
                .sorted((s1, s2) -> Double.compare(s2.getValue(), s1.getValue()))
                .limit(4)
                .forEach(x -> {
                    pieChartToday.add(new PieEntry(x.getValue().floatValue(), x.getKey().category.title));
                });
        this.sliderChartFragment.setPieChartToday(pieChartToday);

        //LAST_WEEK
        List<BarEntry> pieBarChartLastWeek = new ArrayList<>();
        getSumTransactionLastWeek().forEach(x -> pieBarChartLastWeek.add(new BarEntry(
                        (float)LocalDate.parse(x.getKey()).getDayOfWeek().getValue(),
                        x.getValue().floatValue())
                )
        );
        this.sliderChartFragment.setBarChartLastWeek(pieBarChartLastWeek);

        //date now 2024-05-23 23:34:36
        //this week 20 - 26
        //setBarChartThisWeek
        List<BarEntry> pieBarChartThisWeek = new ArrayList<>();
        getSumTransactionThisWeek()
                .forEach(x -> pieBarChartThisWeek.add(new BarEntry(
                        (float)LocalDate.parse(x.getKey()).getDayOfWeek().getValue(), x.getValue().floatValue())
                ));
        this.sliderChartFragment.setBarChartThisWeek(pieBarChartThisWeek);

    }

    private void updateCategoriesWithTransactions() {
        categoryVM.getCategoriesWithTransactions().observe(this, list -> {
            categoryWithTransactions.clear();
            categoryWithTransactions.addAll(list);
            filterTransactionWithUpdateUI();
        });
    }

    /**
     * @return SimpleEntry, where KEY is the date and VALUE is the sum of all completed transactions. Over the this week.
    * */
    private List<AbstractMap.SimpleEntry<String, Double>> getSumTransactionThisWeek(){
        Calendar startThisWeek = Calendar.getInstance();
        startThisWeek.setTime(MainActivity.nowDate);
        startThisWeek.add(Calendar.DAY_OF_WEEK, -startThisWeek.get(Calendar.DAY_OF_WEEK) + 1);

        Calendar endThisWeek = (Calendar) startThisWeek.clone();
        endThisWeek.add(Calendar.DATE, 7);

        //search for all transactions for the THIS week
        return getSumTransactionForPeriod(startThisWeek.getTime(), endThisWeek.getTime());

    }

    /**
     * @return SimpleEntry, where KEY is the date and VALUE is the sum of all completed transactions. Over the past week.
     * */
    private List<AbstractMap.SimpleEntry<String, Double>>  getSumTransactionLastWeek(){
        Calendar startLastWeek = Calendar.getInstance();//last week 13 - 19
        startLastWeek.setTime(MainActivity.nowDate);
        startLastWeek.add(Calendar.DAY_OF_WEEK, -startLastWeek.get(Calendar.DAY_OF_WEEK) + 1);
        startLastWeek.add(Calendar.DAY_OF_MONTH, -7);

        Calendar endLastWeek = (Calendar) startLastWeek.clone();
        endLastWeek.add(Calendar.DATE, 7);

        //search for all transactions for the LAST week
        return getSumTransactionForPeriod(startLastWeek.getTime(), endLastWeek.getTime());
    }


    /**
     * @param start filtering start date.
     * @param end filtering end date.
     * @return SimpleEntry, where KEY is the date and VALUE is the sum of all completed transactions. For the period specified in the date parameters.
     * */
    private List<AbstractMap.SimpleEntry<String, Double>> getSumTransactionForPeriod(Date start, Date end){
        return toTransactionList().stream().filter(x -> x.type == transactionType && (x.dateTime.before(end) && x.dateTime.after(start)))
                .collect(Collectors.groupingBy(Transaction::getDate))//grouping by date
                .entrySet().stream()//Calculation of the amount on a specific day of the week
                .map(x -> new AbstractMap.SimpleEntry<String, Double>(x.getKey(), x.getValue().stream().mapToDouble(y -> y.sum).sum()))
                .collect(Collectors.toList());
    }

    /**
     * @return all Transaction from List&lt;CategoryWithTransactions&gt;
     * @see HomeFragment#categoryWithTransactions */
    private List<Transaction> toTransactionList() {
        return categoryWithTransactions.stream()
                .map(x -> x.transactions)
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }

    public void setUser(User user) {
        this.user = user;
    }
}

File: WalletTracker/app/src/main/java/com/chtima/wallettracker/fragments/HomeFragment.kt
package com.chtima.wallettracker.fragments

import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import com.chtima.wallettracker.R
import com.chtima.wallettracker.adapters.TransactionAdapter
import com.chtima.wallettracker.components.SwitchTransactionView
import com.chtima.wallettracker.fragments.dialogs.AddTransactionDialogFragment
import com.chtima.wallettracker.models.CategoryWithTransactions
import com.chtima.wallettracker.models.DialogObserver
import com.chtima.wallettracker.models.Transaction
import com.chtima.wallettracker.models.Transaction.TransactionType
import com.chtima.wallettracker.utils.CurrencyUtils
import com.chtima.wallettracker.viewModels.CategoryViewModel
import com.chtima.wallettracker.viewModels.TransactionViewModel
import com.chtima.wallettracker.viewModels.UserViewModel
import com.github.mikephil.charting.data.BarEntry
import com.github.mikephil.charting.data.PieEntry
import java.text.SimpleDateFormat
import java.time.LocalDate
import java.time.ZoneId
import java.util.AbstractMap
import java.util.Calendar
import java.util.Date;
import java.util.Locale

import autodispose2.AutoDispose;
import autodispose2.androidx.lifecycle.AndroidLifecycleScopeProvider;
import com.chtima.wallettracker.db.AppDatabase
import com.chtima.wallettracker.models.User


class HomeFragment private constructor(): Fragment() {

    //ViewModels
    private lateinit var userVM: UserViewModel
    private lateinit var categoryVM: CategoryViewModel
    private lateinit var transactionVM: TransactionViewModel

    private lateinit var user: User

    //UI
    private lateinit var btnBalance: Button
    private lateinit var sliderChartFragment: SliderChartFragment;

    private var transactionType : TransactionType = TransactionType.EXPENSE;
    private val categoryWithTransactions: MutableList<CategoryWithTransactions> = ArrayList()
    private lateinit var transactionAdapter: TransactionAdapter


    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        val v : View = inflater.inflate(R.layout.fragment_home, container, false)

        Log.d("SimpleDateFormatNOW", SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date(System.currentTimeMillis())))
        Log.d("SimpleDateFormatNOW-LONG", System.currentTimeMillis().toString())


        //UI
        v.findViewById<ImageButton>(R.id.btn_add).setOnClickListener{ _ ->
            val dialog = AddTransactionDialogFragment.newInstance()
            dialog.addDialogObserver(object : DialogObserver<Transaction>{
                @SuppressLint("CheckResult")
                override fun onSuccess(result: Transaction) {
                    transactionVM.insert(result)
                        .subscribe({ id ->
                            result.id = id;
                            if (TransactionType.EXPENSE == result.type)
                                user.deductFromBalance(result.sum)
                            else
                                user.addToBalance(result.sum)

                            userVM.update(user).subscribe({
                                Log.i("WW-INFO-C", "OK")
                            }, {
                                e -> Log.e("WW-INFO-E", e.toString())
                            })

                        }, { er ->
                            val b = AppDatabase.isExist(requireContext())
                            Log.e("ERR", er.toString())
                        })

                }

                override fun onCancel() {}
            })
            dialog.show(childFragmentManager, AddTransactionDialogFragment::class.java.name)
        }

        //recyclerView
        transactionAdapter = TransactionAdapter(requireContext(), toTransactionList().toMutableList())
        val recyclerView: RecyclerView = v.findViewById(R.id.transaction_recycle)
        recyclerView.adapter = transactionAdapter

        sliderChartFragment = SliderChartFragment.newInstance()
        childFragmentManager.beginTransaction()
                .replace(R.id.slide_chart_fragment, sliderChartFragment, SliderChartFragment.Companion::class.java.name)
                .commit()

        //SwitchTransactionView
        val switch = v.findViewById<SwitchTransactionView>(R.id.swicher_transaction_type);
        switch.addSwitchTransactionListener(object : SwitchTransactionView.SwitchTransactionListener{
            override fun onChangedSelection(type: TransactionType) {
                transactionType = type
                filterTransactionWithUpdateUI()
            }
        })

        //buttons
        btnBalance = v.findViewById<Button>(R.id.balance_btn);//set user balance to "balance_btn"

        return v
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        userVM = ViewModelProvider(this)[UserViewModel::class]
        categoryVM = ViewModelProvider(this)[CategoryViewModel::class]
        transactionVM = ViewModelProvider(this)[TransactionViewModel::class]

        userVM.getUser().observe(viewLifecycleOwner) { u ->
            user = u
            btnBalance.text = "${user.balance} ${CurrencyUtils.getCurrencyChar(user.currency?:"", requireContext())}"
        }

        //как обновить UI
        categoryVM.getCategoriesWithTransactionsByUser().observe(viewLifecycleOwner){
            categoryWithTransactions.clear()
            categoryWithTransactions.addAll(it)
            filterTransactionWithUpdateUI()
        }
    }


    companion object {
        @JvmStatic
        fun newInstance() = HomeFragment().apply {}
    }

    //Method to filter transactions based on the transaction type and update the UI
    private fun filterTransactionWithUpdateUI(){
        if(categoryWithTransactions.isEmpty()) return

        val transactionsFilterType = categoryWithTransactions
            .flatMap { it.transactions }
            .filter { it.type == transactionType }
            .sortedByDescending { it.dateTime }

        transactionAdapter.updateList(transactionsFilterType)

        if(sliderChartFragment.isResumed){
            updateSliderChartFragment()
        }else{
            sliderChartFragment.lifecycle.addObserver(object: DefaultLifecycleObserver{
                override fun onResume(owner: LifecycleOwner) {
                    super.onResume(owner)
                    updateSliderChartFragment()
                }
            })
        }

    }

    private fun updateSliderChartFragment(){
        //setPieChart TODAY
        val pieChartToday: MutableList<PieEntry> = ArrayList()

        categoryWithTransactions.map { item ->
            val sum = item.transactions.stream().filter {
                it.type == transactionType && it.dateTime.toInstant()
                    .atZone(ZoneId.systemDefault())
                    .toLocalDate().equals(
                        Date(System.currentTimeMillis()).toInstant().atZone(ZoneId.systemDefault()).toLocalDate()
                    ) }?.mapToDouble { it.sum }?.sum()
            AbstractMap.SimpleEntry<CategoryWithTransactions, Double>(item, sum)
        }.filter { it.value > 0 }.sortedBy { it.value }.take(4)
            .forEach {
                pieChartToday.add(PieEntry(it.value.toFloat(), it.key.category.title))
            }

        sliderChartFragment.setPieChartToday(pieChartToday)

        //LAST_WEEK
        val pieBarChartLastWeek = ArrayList<BarEntry>()
        getSumTransactionLastWeek().forEach {
            pieBarChartLastWeek.add(BarEntry(LocalDate.parse(it.key).dayOfWeek.value.toFloat(), it.value.toFloat()))
        }
        this.sliderChartFragment.setBarChartLastWeek(pieBarChartLastWeek);

        //setBarChartThisWeek
        val pieBarChartThisWeek = ArrayList<BarEntry>()
        getSumTransactionThisWeek().forEach {
            pieBarChartThisWeek.add(BarEntry(LocalDate.parse(it.key).dayOfWeek.value.toFloat(), it.value.toFloat()))
        }
        this.sliderChartFragment.setBarChartThisWeek(pieBarChartThisWeek);
    }

    /**
     * @return SimpleEntry, where KEY is the date and VALUE is the sum of all completed transactions. Over the past week.
     * */
    private fun getSumTransactionLastWeek(): List<AbstractMap.SimpleEntry<String, Double>>{
        val startLastWeek = Calendar.getInstance();//last week 13 - 19
        startLastWeek.setTime(Date(System.currentTimeMillis()));
        startLastWeek.add(Calendar.DAY_OF_WEEK, -startLastWeek.get(Calendar.DAY_OF_WEEK) + 1);
        startLastWeek.add(Calendar.DAY_OF_MONTH, -7);

        val endLastWeek = startLastWeek.clone() as Calendar
        endLastWeek.add(Calendar.DATE, 7);

        //search for all transactions for the LAST week
        return getSumTransactionForPeriod(startLastWeek.time, endLastWeek.time);
    }

    /**
     * @return SimpleEntry, where KEY is the date and VALUE is the sum of all completed transactions. Over the this week.
     * */
    private fun getSumTransactionThisWeek(): List<AbstractMap.SimpleEntry<String, Double>>{
        val startThisWeek = Calendar.getInstance()
        startThisWeek.setTime(Date(System.currentTimeMillis()))
        startThisWeek.add(Calendar.DAY_OF_WEEK, -startThisWeek.get(Calendar.DAY_OF_WEEK) + 1)

        val endThisWeek = startThisWeek.clone() as Calendar
        endThisWeek.add(Calendar.DATE, 7)

        //search for all transactions for the THIS week
        return getSumTransactionForPeriod(startThisWeek.time, endThisWeek.time);
    }

    /**
     * @param start filtering start date.
     * @param end filtering end date.
     * @return SimpleEntry, where KEY is the date and VALUE is the sum of all completed transactions. For the period specified in the date parameters.
     * */
    private fun getSumTransactionForPeriod(start:Date, end: Date) :  List<AbstractMap.SimpleEntry<String, Double>>{
        return toTransactionList().filter { it.type == transactionType && (it.dateTime.before(end) && it.dateTime.after(start)) }
            .groupBy { it.getDate() }
            .map { AbstractMap.SimpleEntry<String, Double>(it.key, it.value.sumOf { transaction: Transaction -> transaction.sum }) }
    }

    /**
     * @return all Transaction from List&lt;CategoryWithTransactions&gt;
     * @see HomeFragment#categoryWithTransactions */
    private fun toTransactionList():List<Transaction>{
        return categoryWithTransactions.flatMap { it.transactions }

    }

}